shell/platform/fuchsia/flutter/tests/fake_session_unittests.cc

```diff
@@ -0,0 +1,459 @@
+                    FakeNode::kDefaultZeroTranslation,
+                    FakeNode::kDefaultZeroAnchor, FakeNode::kIsHitTestable,
+                    FakeNode::kIsSemanticallyVisible),
+          IsEmpty())));
+}
+
+Matcher<FakeSceneGraph> IsEmptySceneGraph() {
+  return FieldsAre(IsEmpty(), IsEmpty(), IsEmpty(), kInvalidFakeResourceId);
+}
+
+MATCHER_P5(IsBasicSceneGraph,
+           debug_label,
+           node_label,
+           view_holder_koid,
+           view_ref_control_koid,
+           view_ref_koid,
+           "") {
+  static_assert(std::is_same_v<FakeSceneGraph, std::decay_t<decltype(arg)>>);
+  static_assert(
+      std::is_same_v<std::string, std::decay_t<decltype(debug_label)>>);
+  static_assert(
+      std::is_same_v<std::string, std::decay_t<decltype(node_label)>>);
+  static_assert(
+      std::is_same_v<zx_koid_t, std::decay_t<decltype(view_holder_koid)>>);
+  static_assert(
+      std::is_same_v<zx_koid_t, std::decay_t<decltype(view_ref_control_koid)>>);
+  static_assert(
+      std::is_same_v<zx_koid_t, std::decay_t<decltype(view_ref_koid)>>);
+
+  return ExplainMatchResult(
+      FieldsAre(
+          IsEmpty(),
+          AllOf(SizeIs(2u),
+                Contains(Pair(arg.root_view_id,
+                              Pointee(FieldsAre(
+                                  arg.root_view_id, "",
+                                  FakeResource::kDefaultEmptyEventMask,
+                                  VariantWith<FakeView>(FieldsAre(
+                                      view_holder_koid, view_ref_control_koid,
+                                      view_ref_koid, debug_label,
+                                      ElementsAre(Pointee(IsEntityNode(
+                                          _, node_label, IsEmpty()))),
+                                      FakeView::kDebugBoundsDisbaled))))))),
+          _, AllOf(Not(kInvalidFakeResourceId), arg.root_view_id)),
+      arg, result_listener);
+}
+
+}  // namespace
+
+class FakeSessionTest : public ::testing::Test,
+                        public fuchsia::ui::scenic::SessionListener {
+ protected:
+  FakeSessionTest()
+      : session_listener_(this), session_subloop_(loop_.StartNewLoop()) {}
+  ~FakeSessionTest() override = default;
+
+  async::TestLoop& loop() { return loop_; }
+
+  FakeSession& fake_session() { return fake_session_; }
+
+  scenic::Session CreateSession() {
+    FML_CHECK(!fake_session_.is_bound());
+    FML_CHECK(!session_listener_.is_bound());
+
+    auto [session, session_listener] =
+        fake_session_.Bind(session_subloop_->dispatcher());
+    session_listener_.Bind(std::move(session_listener));
+
+    return scenic::Session(session.Bind());
+  }
+
+ private:
+  // |fuchsia::ui::scenic::SessionListener|
+  void OnScenicError(std::string error) override { FML_CHECK(false); }
+
+  // |fuchsia::ui::scenic::SessionListener|
+  void OnScenicEvent(std::vector<fuchsia::ui::scenic::Event> events) override {
+    FML_CHECK(false);
```

Comment:
Here and OnScencError, maybe say
FML_UNREACHABLE() << "unimplemented";
as a hint